### PR TITLE
#1 - Adding Package to Repository

### PR DESCRIPTION
**Pull request by olegmoz**
Add Repository.add() method, which could be used to add packages to repository

### CODE REVIEW

**Comment by g4s8** [comment]
@0crat assign @paulodamaso

**Comment by 0crat** [comment]
> @0crat assign @paulodamaso  ([here](https://github.com/artipie/composer-adapter/issues/18#issuecomment-584700589))

@g4s8 This pull request [#18](https://github.com/artipie/composer-adapter/issues/18) is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/249420e9-fd28-4bd1-8fc2-3bcd25b1c9be); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/test/java/com/artpie/composer/RepositoryTest.java [hunk 1]
```diff
@@ -50,22 +61,27 @@
...
+     */
+    private Package pack;
+
     @BeforeEach
-    void init(final @TempDir Path temp) {
+    void init(final @TempDir Path temp) throws Exception {
         this.storage = new FileStorage(temp);
+        this.pack = new JsonPackage(
+            ByteSource.wrap(
+                ByteStreams.toByteArray(new ResourceOf("minimal-package.json").stream())
```
Comment (paulodamaso) [inline src/test/java/com/artpie/composer/RepositoryTest.java]: @olegmoz Let's use a constant for this String value

File: src/test/java/com/artpie/composer/RepositoryTest.java [hunk 2]
```diff
@@ -50,22 +61,27 @@
...
         final Name name = new Name("foo/bar");
         final Packages packages = new Repository(this.storage).packages(name);
         packages.save(this.storage, name.key()).get();
-        final JsonObject saved;
-        final byte[] bytes = new BlockingStorage(this.storage).value(name.key());
-        try (JsonReader reader = Json.createReader(new ByteArrayInputStream(bytes))) {
-            saved = reader.readObject();
-        }
-        MatcherAssert.assertThat(saved.getJsonObject("packages").keySet(), Matchers.empty());
+        MatcherAssert.assertThat(this.packages(name).keySet(), Matchers.empty());
```
Comment (paulodamaso) [inline src/test/java/com/artpie/composer/RepositoryTest.java]: @olegmoz Regarding to hamcrest `Matchers`, please observe:
- do not use static ones. Prefer to find some matcher object so the code becomes more OO. In this case could be `IsEmptyCollection` I think
- add a message to your assertions, so whoever is executing the tests can see the failure description

File: src/test/java/com/artpie/composer/RepositoryTest.java [hunk 3]
```diff
@@ -80,4 +96,52 @@ void shouldLoadNonEmptyPackages() throws Exception {
...
     }
+
+    @Test
+    void shouldAddPackage() throws Exception {
+        final Key.From key = this.savePackage();
+        new Repository(this.storage).add(key).get();
+        final Name name = this.pack.name();
+        MatcherAssert.assertThat(
+            this.packages(name).getJsonObject(name.string()).keySet(),
+            Matchers.equalTo(Collections.singleton(this.pack.version()))
```
Comment (paulodamaso) [inline src/test/java/com/artpie/composer/RepositoryTest.java]: @olegmoz Please take a look at the comment above

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

File: src/test/java/com/artpie/composer/RepositoryTest.java [hunk 4]
```diff
@@ -80,4 +96,52 @@ void shouldLoadNonEmptyPackages() throws Exception {
...
+        final Name name = this.pack.name();
+        new BlockingStorage(this.storage).save(
+            name.key(),
+            "{\"packages\":{\"vendor/package\":{\"1.1.0\":{}}}}".getBytes()
+        );
+        final Key.From key = this.savePackage();
+        new Repository(this.storage).add(key).get();
+        MatcherAssert.assertThat(
+            this.packages(name).getJsonObject(name.string()).keySet(),
+            Matchers.equalTo(new HashSet<>(Arrays.asList("1.1.0", this.pack.version())))
```
Comment (paulodamaso) [inline src/test/java/com/artpie/composer/RepositoryTest.java]: @olegmoz Please see the comment above, about matchers. Also, why not using `cactoos` `ListOf`?

File: src/test/java/com/artpie/composer/RepositoryTest.java [hunk 5]
```diff
@@ -80,4 +96,52 @@ void shouldLoadNonEmptyPackages() throws Exception {
...
+        );
+    }
+
+    private JsonObject packages(final Name name) {
+        final JsonObject saved;
+        final byte[] bytes = new BlockingStorage(this.storage).value(name.key());
+        try (JsonReader reader = Json.createReader(new ByteArrayInputStream(bytes))) {
+            saved = reader.readObject();
+        }
+        return saved.getJsonObject("packages");
```
Comment (paulodamaso) [inline src/test/java/com/artpie/composer/RepositoryTest.java]: @olegmoz This one could be a constant

File: src/test/java/com/artpie/composer/RepositoryTest.java [hunk 6]
```diff
@@ -80,4 +96,52 @@ void shouldLoadNonEmptyPackages() throws Exception {
...
+    }
+
+    private Key.From savePackage() throws IOException {
+        final byte[] bytes;
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream();
+            JsonWriter writer = Json.createWriter(out)) {
```
Comment (paulodamaso) [inline src/test/java/com/artpie/composer/RepositoryTest.java]: @olegmoz `final`;..

File: src/test/java/com/artpie/composer/RepositoryTest.java (hunk 1 above)
Comment (olegmoz) [reply src/test/java/com/artpie/composer/RepositoryTest.java]: @paulodamaso do not think that it makes sense to extract constant that is used only once and not going to be reused

File: src/test/java/com/artpie/composer/RepositoryTest.java (hunk 2 above)
Comment (olegmoz) [reply src/test/java/com/artpie/composer/RepositoryTest.java]: @paulodamaso fixed

File: src/test/java/com/artpie/composer/RepositoryTest.java (hunk 3 above)
Comment (olegmoz) [reply src/test/java/com/artpie/composer/RepositoryTest.java]: @paulodamaso fixed

File: src/test/java/com/artpie/composer/RepositoryTest.java (hunk 4 above)
Comment (olegmoz) [reply src/test/java/com/artpie/composer/RepositoryTest.java]: @paulodamaso fixed the matcher. I prefer to not use any libraries where java embedded library works just fine

File: src/test/java/com/artpie/composer/RepositoryTest.java (hunk 5 above)
Comment (olegmoz) [reply src/test/java/com/artpie/composer/RepositoryTest.java]: @paulodamaso do not think that it makes sense to extract constant that is used only once and not going to be reused

File: src/test/java/com/artpie/composer/RepositoryTest.java (hunk 6 above)
Comment (olegmoz) [reply src/test/java/com/artpie/composer/RepositoryTest.java]: @paulodamaso sorry, but `qualice` static code analysis tool is against redundant `final` keywords, so `final` is not allowed in try with resource

**Comment by olegmoz** [comment]
@paulodamaso made changes to this PR by your comments, please review

File: src/test/java/com/artpie/composer/RepositoryTest.java (hunk 4 above)
Comment (paulodamaso) [reply src/test/java/com/artpie/composer/RepositoryTest.java]: @olegmoz You can do this more elegantly, readable and Object Oriented using `cactoos`

File: src/test/java/com/artpie/composer/RepositoryTest.java (hunk 1 above)
Comment (paulodamaso) [reply src/test/java/com/artpie/composer/RepositoryTest.java]: Ok

File: src/test/java/com/artpie/composer/RepositoryTest.java (hunk 3 above)
Comment (paulodamaso) [reply src/test/java/com/artpie/composer/RepositoryTest.java]: Thanks!

File: src/test/java/com/artpie/composer/RepositoryTest.java (hunk 5 above)
Comment (paulodamaso) [reply src/test/java/com/artpie/composer/RepositoryTest.java]: @olegmoz OK!

File: src/test/java/com/artpie/composer/RepositoryTest.java (hunk 6 above)
Comment (paulodamaso) [reply src/test/java/com/artpie/composer/RepositoryTest.java]: @olegmoz Ok!

File: src/test/java/com/artpie/composer/RepositoryTest.java (hunk 2 above)
Comment (paulodamaso) [reply src/test/java/com/artpie/composer/RepositoryTest.java]: @olegmoz Thanks!

File: src/test/java/com/artpie/composer/RepositoryTest.java (hunk 4 above)
Comment (olegmoz) [reply src/test/java/com/artpie/composer/RepositoryTest.java]: @paulodamaso fixed

**Comment by olegmoz** [comment]
@paulodamaso please review latest changes

File: src/test/java/com/artpie/composer/RepositoryTest.java [hunk 7]
```diff
@@ -76,8 +96,60 @@ void shouldLoadNonEmptyPackages() throws Exception {
...
+
+    @Test
+    void shouldAddPackage() throws Exception {
+        final Key.From key = this.savePackage();
+        new Repository(this.storage).add(key).get();
+        final Name name = this.pack.name();
+        MatcherAssert.assertThat(
+            "Package with correct version should present in packages after being added",
+            this.packages(name).getJsonObject(name.string()).keySet(),
+            new IsEqual<>(Collections.singleton(this.pack.version()))
```
Comment (paulodamaso) [inline src/test/java/com/artpie/composer/RepositoryTest.java]: @olegmoz I think you could use `SetOf` here too

Comment (olegmoz) [reply src/test/java/com/artpie/composer/RepositoryTest.java]: @paulodamaso thank you, fixed

**Comment by olegmoz** [comment]
@paulodamaso please review

**Comment by paulodamaso** [comment]
@g4s8 please merge

**Comment by olegmoz** [comment]
@yegor256 please review

**Comment by g4s8** [comment]
@olegmoz please merge master branch here and update PR description

**Comment by olegmoz** [comment]
@g4s8 added PR description and updated the branch

**Comment by 0crat** [comment]
Job was finished in 41 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
Order was finished: +20 point(s) just awarded to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Pull request merged**